### ISSUE TITLE
Text edits from efm-langserver are applied in reverse

### ISSUE THREAD

**Issue by kyoh86** [neovim/neovim#29392]
### Problem

Formatting some continuous lines with efm-langserver, text lines are reversed.

It depends on the commit: https://github.com/neovim/neovim/commit/2ce4a4d91e4abee0aab8b98c47eea9fbd4849ba6 and the issue: #29202.

Maybe the changes in the commit is intended to conform to specifications of LSP, but they looks like not still enough.
(Specs doesn't require sorting of text-edits)
So I don't know that which product (efm, Neovim or VSCode?) should fix the bug.

#### Prepare environment

We needs nvim-lspconfig, efm-langserver and jq for example (in Linux):

```console
$ export NVIM_APPNAME=nvim_issue_29392
$ NVIM_APPDIR="$HOME/.config/$NVIM_APPNAME"
$ EFM_NAME=efm-langserver_v0.0.53_linux_amd64
$ JQ_NAME=jq-linux-amd64 

$ mkdir -p "$NVIM_APPDIR/bin"
$ git clone https://github.com/neovim/nvim-lspconfig "$NVIM_APPDIR/rtp/nvim-lspconfig"
$ curl -Lo "$NVIM_APPDIR/bin/efm.tar.gz" "https://github.com/mattn/efm-langserver/releases/download/v0.0.53/$EFM_NAME.tar.gz"
$ curl -Lo "$NVIM_APPDIR/bin/jq" "https://github.com/jqlang/jq/releases/download/jq-1.7.1/$JQ_NAME"
$ chmod +x "$NVIM_APPDIR/bin/jq"
$ cd "$NVIM_APPDIR/bin"
$ tar -xvzf efm.tar.gz
$ mv $EFM_NAME/efm-langserver .
```

#### Setup neovim

And set-up nvim like below (`~/.config/nvim_issue_29392/init.lua`):

```
vim.opt.rtp:append("~/.config/nvim_issue_29392/rtp/nvim-lspconfig")

local lspconfig = require("lspconfig")
lspconfig.efm.setup({
  init_options = {
    documentFormatting = true,
  },
  settings = {
    languages = {
      json = { {
        formatStdin = true,
        formatCommand = "jq .",
      } },
    },
  },
})
```

#### Reproduce in Neovim

Prepare target file:

```json:foo.json
  {
    "foo": "bar"
  }
```

Start neovim:

```console
$ export NVIM_APPNAME=nvim_issue_29392
$ export PATH="$HOME/.config/$NVIM_APPNAME/bin:$PATH"
$ nvim foo.json
```

Call `lua vim.lsp.buf.format()`.

And get the invalid result:

```
}
  "foo": "bar"
{
```

### Expected behavior

Formatted right:

```json
{
  "foo": "bar"
}
```

### Neovim version (nvim -v)

v0.11.0-dev-b0c336e

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Arch Linux

### Terminal name/version

WezTerm

### $TERM environment variable

xterm-256color

### Installation

Build from repo

**Comment by scohen** [comment]
> Maybe the changes in the commit is intended to conform to specifications of LSP, but they looks like not still enough.
(Specs doesn't require sorting of text-edits)

For what it's worth, the spec specifically says that the order of edits matters. The language server I run sends them in a specific order, and they work fine in VSCode. Neovim is the only client that we've had this issue with.

**Comment by kyoh86** [comment]
Just to be clear: I am not sure if the problem @scohen is facing is the same as the one caused by the compatibility with efm-langserver.

**Comment by scohen** [comment]
My take is that the client shouldn't be reordering edits, neovim's client does. Right now, my LS doesn't have a problem with neovim, the commits mentioned in the bug fixed the problems we were having, but I felt that was not a correct fix.

**Comment by kyoh86** [comment]
Maybe it should be reported as another issue.

**Comment by geril2207** [comment]
Also I have kind of the same problem with tsserver after https://github.com/neovim/neovim/commit/2ce4a4d91e4abee0aab8b98c47eea9fbd4849ba6, https://github.com/neovim/neovim/issues/29286

**Comment by alcolmenar** [comment]
@kyoh86 the PR that was landed causing this issue actually is more inline with the LSP spec. The commit makes it so that text edits with the same start position aren't applied in reverse anymore and are applied in the order that they are received.

For text edits that have different start positions, from the LSP spec, they should be provided in reverse order, i.e. edits provided starting from the bottom of the file. If they aren't provided in reverse order, the sorting enforces that. Again, this doesn't apply to text edits with the same start position.

I was worried that this might cause issues because the bug was introduced close to 3 years ago and LSPs that were developed specifically for Neovim and tested against the errant code would have issues.

If you provide the debug logs for your particular case, I can check to see it is relying on the errant code. 

I think we'd need @mfussenegger or @hrsh7th to weigh in on whether the "fix" should be reverted or if LSPs relying on the errant code should be forced to update to be more inline with spec. It's a bit of a tough call, but I lean toward the latter.

**Comment by kyoh86** [comment]
@alcolmenar 
I am not sure if I understand exactly what you are calling "Debug Log", but I set the LSP log level to DEBUG and logged it. How does it look?
```
[START][2024-07-05 01:33:53] LSP logging initiated
[ERROR][2024-07-05 01:33:53] .../vim/lsp/rpc.lua:772	"rpc"	"/home/kyoh86/.config/nvim_issue_29392/bin/efm-langserver"	"stderr"	"2024/07/05 01:33:53 efm-langserver: no configuration file\n2024/07/05 01:33:53 efm-langserver: reading on stdin, writing on stdout\n"
[DEBUG][2024-07-05 01:35:00] ...m/lsp/client.lua:678	"LSP[efm]"	"client.request"	1	"textDocument/formatting"	{  options = {    insertSpaces = false,    tabSize = 8  },  textDocument = {    uri = "file:///home/kyoh86/foo.json"  }}	<function 1>	1
[DEBUG][2024-07-05 01:35:00] .../vim/lsp/rpc.lua:286	"rpc.send"	{  id = 2,  jsonrpc = "2.0",  method = "textDocument/formatting",  params = {    options = {      insertSpaces = false,      tabSize = 8    },    textDocument = {      uri = "file:///home/kyoh86/foo.json"    }  }}
[DEBUG][2024-07-05 01:35:00] .../vim/lsp/rpc.lua:408	"rpc.receive"	{  id = 2,  jsonrpc = "2.0",  result = { {      newText = "",      range = {        ["end"] = {          character = 0,          line = 3        },        start = {          character = 0,          line = 0        }      }    }, {      newText = "{\n",      range = {        ["end"] = {          character = 0,          line = 3        },        start = {          character = 0,          line = 3        }      }    }, {      newText = '  "foo": "bar"\n',      range = {        ["end"] = {          character = 0,          line = 3        },        start = {          character = 0,          line = 3        }      }    }, {      newText = "}\n",      range = {        ["end"] = {          character = 0,          line = 3        },        start = {          character = 0,          line = 3        }      }    } }}
[DEBUG][2024-07-05 01:35:00] .../vim/lsp/rpc.lua:286	"rpc.send"	{  jsonrpc = "2.0",  method = "textDocument/didChange",  params = {    contentChanges = { {        text = '}\n  "foo": "bar"\n{\n\n'      } },    textDocument = {      uri = "file:///home/kyoh86/foo.json",      version = 7    }  }}
[INFO][2024-07-05 01:35:03] .../lua/vim/lsp.lua:804	"exit_handler"	{ {    _log_prefix = "LSP[efm]",    _on_attach_cbs = { <function 1> },    _on_exit_cbs = { <function 2>, <function 3> },    _on_init_cbs = { <function 4> },    _trace = "off",    attached_buffers = { true },    cancel_request = <function 5>,    capabilities = <1>{      general = {        positionEncodings = { "utf-16" }      },      textDocument = {        callHierarchy = {          dynamicRegistration = false        },        codeAction = {          codeActionLiteralSupport = {            codeActionKind = {              valueSet = { "", "quickfix", "refactor", "refactor.extract", "refactor.inline", "refactor.rewrite", "source", "source.organizeImports" }            }          },          dataSupport = true,          dynamicRegistration = true,          isPreferredSupport = true,          resolveSupport = {            properties = { "edit" }          }        },        completion = {          completionItem = {            commitCharactersSupport = false,            deprecatedSupport = false,            documentationFormat = { "markdown", "plaintext" },            preselectSupport = false,            resolveSupport = {              properties = { "additionalTextEdits" }            },            snippetSupport = true          },          completionItemKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 }          },          completionList = {            itemDefaults = { "editRange", "insertTextFormat", "insertTextMode", "data" }          },          contextSupport = false,          dynamicRegistration = false        },        declaration = {          linkSupport = true        },        definition = {          dynamicRegistration = true,          linkSupport = true        },        diagnostic = {          dynamicRegistration = false        },        documentHighlight = {          dynamicRegistration = false        },        documentSymbol = {          dynamicRegistration = false,          hierarchicalDocumentSymbolSupport = true,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }          }        },        formatting = {          dynamicRegistration = true        },        hover = {          contentFormat = { "markdown", "plaintext" },          dynamicRegistration = true        },        implementation = {          linkSupport = true        },        inlayHint = {          dynamicRegistration = true,          resolveSupport = {            properties = { "textEdits", "tooltip", "location", "command" }          }        },        publishDiagnostics = {          dataSupport = true,          relatedInformation = true,          tagSupport = {            valueSet = { 1, 2 }          }        },        rangeFormatting = {          dynamicRegistration = true,          rangesSupport = true        },        references = {          dynamicRegistration = false        },        rename = {          dynamicRegistration = true,          prepareSupport = true        },        semanticTokens = {          augmentsSyntaxTokens = true,          dynamicRegistration = false,          formats = { "relative" },          multilineTokenSupport = false,          overlappingTokenSupport = true,          requests = {            full = {              delta = true            },            range = false          },          serverCancelSupport = false,          tokenModifiers = { "declaration", "definition", "readonly", "static", "deprecated", "abstract", "async", "modification", "documentation", "defaultLibrary" },          tokenTypes = { "namespace", "type", "class", "enum", "interface", "struct", "typeParameter", "parameter", "variable", "property", "enumMember", "event", "function", "method", "macro", "keyword", "modifier", "comment", "string", "number", "regexp", "operator", "decorator" }        },        signatureHelp = {          dynamicRegistration = false,          signatureInformation = {            activeParameterSupport = true,            documentationFormat = { "markdown", "plaintext" },            parameterInformation = {              labelOffsetSupport = true            }          }        },        synchronization = {          didSave = true,          dynamicRegistration = false,          willSave = true,          willSaveWaitUntil = true        },        typeDefinition = {          linkSupport = true        }      },      window = {        showDocument = {          support = true        },        showMessage = {          messageActionItem = {            additionalPropertiesSupport = false          }        },        workDoneProgress = true      },      workspace = {        applyEdit = true,        configuration = true,        didChangeConfiguration = {          dynamicRegistration = false        },        didChangeWatchedFiles = {          dynamicRegistration = false,          relativePatternSupport = true        },        inlayHint = {          refreshSupport = true        },        semanticTokens = {          refreshSupport = true        },        symbol = {          dynamicRegistration = false,          symbolKind = {            valueSet = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26 }          }        },        workspaceEdit = {          resourceOperations = { "rename", "create", "delete" }        },        workspaceFolders = true      }    },    commands = {},    config = {      autostart = true,      capabilities = <table 1>,      cmd = { "/home/kyoh86/.config/nvim_issue_29392/bin/efm-langserver" },      cmd_cwd = "/home/kyoh86",      handlers = <2>{},      init_options = {        documentFormatting = true      },      log_level = 2,      message_level = 2,      name = "efm",      on_attach = <function 1>,      on_exit = <function 2>,      on_init = <function 4>,      settings = <3>{        languages = {          json = { {              formatCommand = "jq .",              formatStdin = true            } }        }      },      single_file_support = true,      <metatable> = {        __tostring = <function 6>      }    },    dynamic_capabilities = {      capabilities = {},      client_id = 1,      <metatable> = {        __index = {          get = <function 7>,          match = <function 8>,          new = <function 9>,          register = <function 10>,          supports = <function 11>,          supports_registration = <function 12>,          unregister = <function 13>        }      }    },    flags = {},    get_language_id = <function 14>,    handlers = <table 2>,    id = 1,    initialized = true,    is_stopped = <function 15>,    messages = {      messages = {},      name = "efm",      progress = {},      status = {}    },    name = "efm",    notify = <function 16>,    offset_encoding = "utf-16",    on_attach = <function 17>,    progress = {      _idx_read = 0,      _idx_write = 0,      _items = {},      _size = 51,      pending = {},      <metatable> = {        __call = <function 18>,        __index = {          clear = <function 19>,          peek = <function 20>,          pop = <function 21>,          push = <function 22>        }      }    },    request = <function 23>,    request_sync = <function 24>,    requests = {},    rpc = {      is_closing = <function 25>,      notify = <function 26>,      request = <function 27>,      terminate = <function 28>    },    server_capabilities = {      documentFormattingProvider = true,      textDocumentSync = {        change = 1,        openClose = true,        save = {          includeText = false        },        willSave = false,        willSaveWaitUntil = false      },      workspace = {        workspaceFolders = {          changeNotifications = true,          supported = true        }      }    },    settings = <table 3>,    stop = <function 29>,    supports_method = <function 30>,    workspace_did_change_configuration = <function 31>,    <metatable> = <4>{      __index = <table 4>,      _add_workspace_folder = <function 32>,      _cancel_request = <function 33>,      _exec_cmd = <function 34>,      _is_stopped = <function 35>,      _notification = <function 36>,      _notify = <function 37>,      _on_attach = <function 38>,      _on_error = <function 39>,      _on_exit = <function 40>,      _remove_workspace_folder = <function 41>,      _request = <function 42>,      _request_sync = <function 43>,      _resolve_handler = <function 44>,      _run_callbacks = <function 45>,      _server_request = <function 46>,      _stop = <function 47>,      _supports_method = <function 48>,      _text_document_did_open_handler = <function 49>,      create = <function 50>,      initialize = <function 51>,      write_error = <function 52>    }  } }
[DEBUG][2024-07-05 01:35:03] .../vim/lsp/rpc.lua:286	"rpc.send"	{  id = 3,  jsonrpc = "2.0",  method = "shutdown"}
[ERROR][2024-07-05 01:35:03] .../vim/lsp/rpc.lua:772	"rpc"	"/home/kyoh86/.config/nvim_issue_29392/bin/efm-langserver"	"stderr"	"2024/07/05 01:35:03 jsonrpc2 handler: sending response 3: jsonrpc2: connection is closed\n2024/07/05 01:35:03 efm-langserver: connections closed\n"
```

**Comment by kyoh86** [comment]
> I was worried that this might cause issues because the bug was introduced close to 3 years ago and LSPs that were developed in that timeframe specifically for Neovim and tested against the errant code would have issues.

As you say, and quite sadly, the specification has already been rebuffed by Microsoft (in tsserver, for example), the party that formulated it.
But even if it is the "correct" implementation of the specification, if users will get destructive results, I think we should try to cooperate with LS to avoid such a tragedy.
Of course, I understand that the effort involved is significant, and I am not going to ask for it unilaterally. I would be willing to request tsserver to correct this discrepancy myself.

**Comment by alcolmenar** [comment]
> @alcolmenar I am not sure if I understand exactly what you are calling "Debug Log", but I set the LSP log level to DEBUG and logged it. How does it look?

Looks good and thanks. It does appear that the LSP is leveraging the errant code. If we look at what's being received from the LSP:

```lua
{
	id = 2,
	jsonrpc = "2.0",
	result = {
		{
			newText = "",
			range = {
				["end"] = { character = 0, line = 3 },
				start = { character = 0, line = 0 },
			},
		},
		{
			newText = "{\n",
			range = {
				["end"] = { character = 0, line = 3 },
				start = { character = 0, line = 3 },
			},
		},
		{
			newText = '  "foo": "bar"\n',
			range = {
				["end"] = { character = 0, line = 3 },
				start = { character = 0, line = 3 },
			},
		},
		{
			newText = "}\n",
			range = {
				["end"] = { character = 0, line = 3 },
				start = { character = 0, line = 3 },
			},
		},
	},
}
```

The result field is an array of text edits. Elements 2-4 all have the same start position and with the "fix", these aren't being applied in reverse order anymore. Because they have the same start position and the same end position, the `newText` is inserted at that position without any overwrite. This results in the issue that you're seeing. Hope that makes sense

**Comment by mfussenegger** [comment]
The response looks like the text edit starting on line=3 all depend on the first text edit being applied first (line=0-3).

So they move the document from:

S1:

```json
  {
    "foo": "bar"
  }
```

to:

```json
```

to:

```json
{
```

```json
{
    "foo": "bar"
```

to:

```json
{
    "foo": "bar"
}

```


But if I understand the specification right, that's wrong as _all_ text edits must refer to positions relative to the original state, and all together move from S1 to S2. That means the second text edit can't rely on the first text edit.

> All text edits ranges refer to positions in the document they are computed on. They therefore move a document from state S1 to S2 without describing any intermediate state. Text edits ranges must never overlap, that means no part of the original document must be manipulated by more than one edit. However, it is possible that multiple edits have the same start position: multiple inserts, or any number of inserts followed by a single remove or replace edit. If multiple inserts have the same position, the order in the array defines the order in which the inserted strings appear in the resulting text.


Note that for the editor to move from S1 to S2 without interference it must apply the text edits in reverse order, (with the exception of the same starting line, which the linked commit fixed).

This also was the intention of one of the original commits based by the comments: https://github.com/neovim/neovim/commit/bcae04f6c62b23104e85bb08c54a16d0cdc33852

**Comment by kyoh86** [comment]
@alcolmenar I see. thanks!
I'll try to report to efm-langserver repository.

**Comment by kyoh86** [comment]
NOTE:
Vim (with [vim-lsp](https://github.com/prabirshrestha/vim-lsp)) and [Helix](https://github.com/helix-editor/helix) doesn't applies them in reverse.
Maybe VSCode also.
We need to request such other clients to fix the order, like  https://github.com/neovim/neovim/commit/2ce4a4d91e4abee0aab8b98c47eea9fbd4849ba6 and the issue https://github.com/neovim/neovim/issues/29202.

It's going to be a very long road...

**Comment by kyoh86** [comment]
coc.nvim also doesn't applies them in reverse.
Now only the Neovim is buggy. (But it's as per specifications)

**Comment by kyoh86** [comment]
I reported other clients:

https://github.com/helix-editor/helix/issues/11174
https://github.com/neoclide/coc.nvim/issues/5087
https://github.com/prabirshrestha/vim-lsp/issues/1566

I'm waiting their response.

**Comment by kyoh86** [comment]
helix and coc.nvim follows how VScode does, not specification.

It seems that no one but Neovim intends to follow this specification. As it is, it seems that the various language servers have no choice but to separate the responses for Neovim from the responses for the others.
It would be very silly...

@alcolmenar @justinmk 
Is conformance to this specification more important than cooperation with other editors and the increased burden on many language server implementors?

**Comment by mfussenegger** [comment]
> Is conformance to this specification more important 

Yes. Otherwise it devalues the specification and it turns into wild west where everyone has to follow the vscode implementation while the specification turns into mere decoration. 

If vscode doesn't follow the specification that should be brought up with them and they must change it if they want to be compliant. If they refuse they should adapt the specification - given that Microsoft controls both that shouldn't be an issue.

**Comment by kyoh86** [comment]
@mfussenegger 

Thanks for the quick response.
I would like to raise this issue with the VSCode project, as discussing only about this project will not solve the problem.
However, I am not bright about VSCode's LSP implementation.
Also, I could not set up conditions that would reproduce this issue.
Can anyone help me to petition the VSCode side to rectify this situation?
I am very sorry for my lack of help.

**Comment by mfussenegger** [comment]
Okay, I had a closer look at this again. I think vscode actually matches the spec, and the fix we merged was wrong. See https://github.com/neovim/neovim/pull/29877

**Comment by kyoh86** [comment]
Thanks! it's fixed in bdff50d
